Web application responsive & color selector | Code clean

gottburgm opened the pull request:
### **General** 
  - Cleaned as much as i could the html/css and JavaScript hardcoded stuff. I moved all the JavaScript code in **wardemo/src/main/webapp/js/script.js** which is included by the html.
  - Added responsive HTML/CSS support to fit on mobile devices .
  - Added a color selector which allow the user(s) to choose a color and a led to fill with this color from the web application .
 - **WANRNING:** The HTML/CSS or JavaScript may contain small errors as i couldn't test the last changes that i made on the raspberry. It is working on my browser but need to be compiled and tested on a raspberry.

### **Directories** 
  - **wardemo/src/main/webapp/css** for the CSS files .
  - **wardemo/src/main/webapp/js** for the JavaScript files .

### **Code**
  - Added HTML headers to follow W3 rules
  - Added **style.css** for custom CSS rules
  - Added **script.js** which contains all the JavaScript functions
  - Removed hardcoded IP addresses / references
  
  ```java
    // Removed from FrameBuffer.java
   out.println("var request = new XMLHttpRequest();");
   out.println("var theUrl = \"http://10.0.0.203:8080/demo-1.0-SNAPSHOT/FrameBuffer?i=\" + " + i + " + \"&j=\" + " + j + ";");
   out.println("request.open( \"GET\", theUrl, false );");
  ```

  ```javascript
   var client = new XMLHttpRequest();
   var requestUrl = window.location.protocol + "//" + window.location.host + "/" + window.location.pathname + "?i=" + event.target.i + "&j=" + event.target.j + "&red=" + getRed(userColor.value) + "&green=" + getGreen(userColor.value) + "&blue=" + getBlue(userColor.value);

   client.open("GET", requestUrl);
   client.send();
  ```
  
  ```
   The request URL looks like:
   http://<RASPBERRY_IP>/<FILE>?i=<CORD_X>&j=<CORD_Y>&red=<RED_HEX_CODE>&green=<GREEN_HEX_CODE>&blue=<BLUE_HEX_CODE>)

   => The RGB value of the elected color is splitted (R, G and B) and sent with the coords of the pixel clicked (so the pixel on which the color should be applied). The page will return the new "pixel mapping"
  ```


  - Removed all debugging stuff
  - Added the correction of the RGB565 function:
  
  ```java
    // return (short) (blue*256+red*16+green/8+(green%8)*512);
    return (short) ((blue<<8)+red*8+green/8+((green%8)<<13));
  ```

  - Added predefined colors because some colors are impossible to make with the leds of the Raspberry PI. Those colors are defined in **script.js** : 

  ```javascript
   colors: {
     'tetrad1': 'rgb(255, 0, 6)',
     'tetrad2': 'rgb(27, 255, 0)',
     'tetrad3': 'rgb(225, 255, 0)',
     'tetrad4': 'rgb(255, 116, 0)',
     'tetrad5': 'rgb(0, 1, 255)',
     'tetrad6': 'rgb(228, 0, 255)',
     'tetrad7': 'rgb(0, 255, 19)',
     'tetrad8': 'rgb(255, 255, 255)',
     'tetrad9': 'rgb(0, 0, 0)'
   },
  ```

### TODO
  - [ ] Test this commit on the Raspberry PI
  - [ ] Test the "responsivility" from different mobile devices 
  - [ ] Read all the HTML which is hardcoded in **FrameBuffer.java** from a template file
  - [ ] Remove useless things from the JavaScript code
  - [ ] Clean the JavaScript code
  - [ ] Use POST requests instead of GET requests

### **Librairies & References** 
  -  Responsive Design & CSS: [Bootstrap Website](https://getbootstrap.com/)
  - Color selection: 
    - [Bootstrap Colorpicker](https://github.com/farbelous/bootstrap-colorpicker.git)
    - [Bootstrap Colorpicker Examples](https://farbelous.io/bootstrap-colorpicker/tutorial-Basics.html)

### **Contact**
  - Github: [gottburgm](https://github.com/gottburgm/)
  - Email: <a href="mailto:mgottbur@redhat.com">mgottbur@redhat.com</a>

gottburgm committed:
- I cleaned as much as i could the html/css and javascript hardcoded stuff. I moved
    all the javascript code in **wardemo/src/main/webapp/js/script.js** which is included by the html

  - I added responsive html/css support to fit on mobile devices

  - I added a color selector which allow the user(s) to choose a color and a
    led to fill with this color from the web application .

  - The html/css or javascript may contain small errors as i couldn't test the last changes that i made on the raspberry. It is working on my browser but need to be compiled and tested on a raspberry

  - **wardemo/src/main/webapp/css** for the css files
  - **wardemo/src/main/webapp/js** for the javascript files

  - Responsive Design & CSS: [Bootstrap Website](https://getbootstrap.com/)
  - Color selection: [Bootstrap Colorpicker](https://github.com/farbelous/bootstrap-colorpicker.git)

jfclere commented on wardemo/src/main/java/org/jfclere/tomcatPI/FrameBuffer.java:
```diff
@@ -23,114 +23,101 @@
+        if (param_j != null && param_i != null) {
+            int p_i = Integer.parseInt(param_i);
+            int p_j = Integer.parseInt(param_j);
+            int red = Integer.parseInt(param_red);
```
What happens is red, green or blue are null?

gottburgm replied:
Well,  this probably is a better solution but they shouldn't be null at any time as the JavaScript which send the request is triggered only when a color (so a RGB value) is selected.

to be moresecure/clean i would change the following things in the **script.js** : 

  ```javascript

  function getRed(rgb_str) {
    var red = 0;
    var rgb_red = /rgba?\(\s*(\d+),(?:[\d\s]*),(?:[\d\s]*)(?:,\s*[\d\.]+)?\)/.exec(rgb_str);
    if(rgb_red) {
       red = rgb_red;
    }

    return red;
  }

  function getGreen(rgb_str) {
    var green = 0;
    var rgb_green = /rgba?\((?:[\d\s]*),\s*(\d+),(?:[\d\s]*)(?:,\s*[\d\.]+)?\)/.exec(rgb_str);
    if(rgb_green) {
       green = rgb_green;
    }

    return green;
  }

  function getBlue(rgb_str) {
    var blue = 0;
    var rgb_blue = /rgba?\((?:[\d\s]*),(?:[\d\s]*),\s*(\d+)(?:,\s*[\d\.]+)?\)/.exec(rgb_str);
    if(rgb_blue) {
       blue = rgb_blue;
    }

    return blue;
  }


  ```

  and in FrameBuffer.java :
  
  ```java
    if (param_j != null && param_i != null && param_red != null && param_green != null && param_blue != null) {
        int p_i = Integer.parseInt(param_i);
        int p_j = Integer.parseInt(param_j);
        int red = Integer.parseInt(param_red);
        int green = Integer.parseInt(param_green);
        int blue = Integer.parseInt(param_blue);

        if(p_i >= 0 && p_i <= 7 && p_j >= 0 && p_j <= 7 && pi.color(red, green, blue)) {
            pi.writepix(p_i, p_j, pi.color(red, green, blue));
        }
    }
  
  ```

jfclere replied:
+1

jfclere commented on wardemo/src/main/java/org/jfclere/tomcatPI/PIFrameBuffer.java:
```diff
@@ -59,63 +58,11 @@ public int getGreen(short pixel) {
 		return (pixel & cache) * 8 + green;
 	}
-
+	
```
If you remove the code for main() why you didn't remove it?
We don't need it correct?

gottburgm replied:
Was not sure about it, we can remove it !

jfclere replied:
OK Cool remove it and update the PR.

jfclere commented on wardemo/src/main/webapp/css/bootstrap-colorpicker.css:
```diff
@@ -0,0 +1,244 @@
+.colorpicker-saturation {
+  width: 100px;
+  height: 100px;
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAQAAADa613fAAAP9klEQVR4XnRWC47rNgwcKjlA0bv2VL1Qi/YELRav7203iS1ppqZoiXCAhuBHVLI74xFtG3/Hz2joIOjRGuR5eMYuRn9YA1fds859KX8ZvczLr9/pImiR3Rqky9/wlajRIdVE/1Rufeu/0No3/ASgBZAJUkwgi0iCaEatekJJoEqiTQncd67/gyOfRCZshTed0Nl8LbLj8D6qxtoq9/7kJz/aH/3Xfu8VwI5+AUH8DxE7gUyiIpZ5LwiGzUqE3CScJsCDQHAsvBnxWpkbC0QMHmBp6latWS0bnvrCN/x1+xPfce+I ...
```
No sure that is right we need the original png file in the repo.

gottburgm replied:
Well as i didn't edited the file, and it comes directly from the official repository i don't think that we should change it.

jfclere commented on wardemo/src/main/webapp/css/bootstrap-colorpicker.css:
```diff
@@ -0,0 +1,244 @@
+  margin-top: -1px; }
+
+.colorpicker-hue {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAABkCAMAAABw8qpSAAABLFBMVEXqFBb/ABH/ACL/ADH/AEH/AFD/AGD/AG7/AH7/AI3/AJ3/AKz/ALz/AMr/ANv/AOr/APr2AP/mAP/XAP/HAP+4AP+oAP+aAP+JAP97AP9rAP9cAP9MAP8+AP8tAP8fAP8PAP8BAv8AEP8AH/8AL/8APv8ATv8AXP8Abf8Ae/8Ai/8Amv8Aqv8AuP8Ayf8A1/8A5/8A9/8A//gA/+kA/9kA/8oA/7oA/6wA/5sA/40A/30A/24A/14A/1AA/z8A/zEA/yEA/xEB/wMN/wAd/wAs/wA8/wBK/wBb/wBp/wB5 ...
```
Same as before.

gottburgm replied:
Same as before for me too :)

jfclere commented on wardemo/src/main/webapp/css/bootstrap-colorpicker.css:
```diff
@@ -0,0 +1,244 @@
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAABkCAMAAABw8qpSAAABLFBMVEXqFBb/ABH/ACL/ADH/AEH/AFD/AGD/AG7/AH7/AI3/AJ3/AKz/ALz/AMr/ANv/AOr/APr2AP/mAP/XAP/HAP+4AP+oAP+aAP+JAP97AP9rAP9cAP9MAP8+AP8tAP8fAP8PAP8BAv8AEP8AH/8AL/8APv8ATv8AXP8Abf8Ae/8Ai/8Amv8Aqv8AuP8Ayf8A1/8A5/8A9/8A//gA/+kA/9kA/8oA/7oA/6wA/5sA/40A/30A/24A/14A/1AA/z8A/zEA/yEA/xEB/wMN/wAd/wAs/wA8/wBK/wBb/wBp/wB5 ...
+
+.colorpicker-alpha {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAABkCAQAAAAVxWkcAAABr0lEQVR4Xo2VwU0DQQxF7dmRuNIFlzlSAR3QAaXQQdIBJVABFXDcOVAAd67cjJLR07dkhcSrkZKfb/t7bG88rFo3B5gZPMNycItu2xloGV7MWHzM9zuzFWCkmA0nK6AszCUJDW6+mG6R03ncw5v8EMTEvZ2O3AliYjpslblc0RF9LmZYWxURU6aKytWZYsoWCAe+xwOZp1GsEukGiIkYxcQCHck99+gRgB7JncyIB5SGEhP3Yh5P6JwX+u6AnYot104d8DJT7uH7M9JH6OZbimj0vfMVaYnJIZFJDBW9kHlerL2C ...
```
Idem

gottburgm replied:
again .

jfclere commented on wardemo/src/main/webapp/css/bootstrap-colorpicker.css:
```diff
@@ -0,0 +1,244 @@
+  margin-top: 0; }
+
+.colorpicker.colorpicker-horizontal .colorpicker-hue {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAABCAMAAAAfBfuPAAABLFBMVEXqFBb/ABH/ACL/ADH/AEH/AFD/AGD/AG7/AH7/AI3/AJ3/AKz/ALz/AMr/ANv/AOr/APr2AP/mAP/XAP/HAP+4AP+oAP+aAP+JAP97AP9rAP9cAP9MAP8+AP8tAP8fAP8PAP8BAv8AEP8AH/8AL/8APv8ATv8AXP8Abf8Ae/8Ai/8Amv8Aqv8AuP8Ayf8A1/8A5/8A9/8A//gA/+kA/9kA/8oA/7oA/6wA/5sA/40A/30A/24A/14A/1AA/z8A/zEA/yEA/xEB/wMN/wAd/wAs/wA8/wBK/wBb/wBp/wB5 ...
```
sure?

jfclere commented on wardemo/src/main/webapp/css/bootstrap-colorpicker.css:
```diff
@@ -0,0 +1,244 @@
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAABCAMAAAAfBfuPAAABLFBMVEXqFBb/ABH/ACL/ADH/AEH/AFD/AGD/AG7/AH7/AI3/AJ3/AKz/ALz/AMr/ANv/AOr/APr2AP/mAP/XAP/HAP+4AP+oAP+aAP+JAP97AP9rAP9cAP9MAP8+AP8tAP8fAP8PAP8BAv8AEP8AH/8AL/8APv8ATv8AXP8Abf8Ae/8Ai/8Amv8Aqv8AuP8Ayf8A1/8A5/8A9/8A//gA/+kA/9kA/8oA/7oA/6wA/5sA/40A/30A/24A/14A/1AA/z8A/zEA/yEA/xEB/wMN/wAd/wAs/wA8/wBK/wBb/wBp/wB5 ...
+
+.colorpicker.colorpicker-horizontal .colorpicker-alpha {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAAKCAQAAADoFTP1AAAB9ElEQVR4XoWTQW4VMRBEu9qWEimL7DhEMp8NF+ASnJJLcAQgE1bcgBUSkYKUuHCrZ9pjeqSU5Yn9LPu7umJQBIIv+k7vIOrtK66L4lmr3pVOrOv3otp619KZ0/KjdNI79L52Uo09FBQWrU0vfe5trezU+hLsoUKd3Repovte+0vbq/7Lj5XbaHECKasR9G4MPlbp+gzZxd6koPEJCkAYC5SjcOTAIIOK90Dja1IfIZ8Z+zAY9jm3b5Ia+MT5sFcqRJrR2AYYA8Kua5BzYRrFPNmD4PQMegGJMOffJJUsWiI3nCHZ ...
```
sure?

gottburgm replied:
What about you ?

jfclere commented on wardemo/src/main/webapp/css/bootstrap-colorpicker.css:
```diff
@@ -0,0 +1,244 @@
+
+/** EXTENSIONS **/
+.colorpicker-preview {
+  background-image: url("data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAoAAABkCAQAAAAVxWkcAAABr0lEQVR4Xo2VwU0DQQxF7dmRuNIFlzlSAR3QAaXQQdIBJVABFXDcOVAAd67cjJLR07dkhcSrkZKfb/t7bG88rFo3B5gZPMNycItu2xloGV7MWHzM9zuzFWCkmA0nK6AszCUJDW6+mG6R03ncw5v8EMTEvZ2O3AliYjpslblc0RF9LmZYWxURU6aKytWZYsoWCAe+xwOZp1GsEukGiIkYxcQCHck99+gRgB7JncyIB5SGEhP3Yh5P6JwX+u6AnYot104d8DJT7uH7M9JH6OZbimj0vfMVaYnJIZFJDBW9kHlerL2C ...
```
sure?

gottburgm replied:
OKAY: https://cdnjs.cloudflare.com/ajax/libs/bootstrap-colorpicker/2.5.2/css/bootstrap-colorpicker.css

gottburgm committed:
- Used POST requests instead of GET
- Added few tests in the javascript color functions and changed the regexes
- Removed some wrong HTML
- Removed some useless HTML
- Replaced data uri images (base64 encoded inline) by path
- Extracted inline images from css and created their corresponding image files into **wardemo/src/main/webapp/images**
- Added missing tests into the JAVA (null values, ..)
- Removed the empty main from **wardemo/src/main/java/org/jfclere/tomcatPI/PIFrameBuffer.java**

- Created HTML template files
- Created JSP Index file which use the JAVA classes and include html templates
- The JSP File seems to work like a charm (tested on my local tomcat server)

- Need to clean a bit the paths and informations stored in XML files to have a nice servlet !
- Still need to be tested on a Raspberry PI

gottburgm committed:
Added MIT LICENSE

- Added MIT LICENSE

gottburgm committed:
Merge pull request #1 from gottburgm/add-license-1

Added MIT LICENSE

jfclere commented on wardemo/src/main/java/org/jfclere/tomcatPI/FrameBuffer.java:
```diff
@@ -115,9 +112,9 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
+	 * @see HttpServlet#doGet(HttpServletRequest request, HttpServletResponse response)
 	 */
-	protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+	protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
```
Sure you want an infinite loop there?
